sdk/src/main/java/com/mapbox/maps/renderer/widget/BitmapWidgetRenderer.kt

```diff
@@ -0,0 +1,307 @@
+package com.mapbox.maps.renderer.widget
+
+import android.graphics.Bitmap
+import android.opengl.GLES20
+import android.opengl.GLUtils
+import android.opengl.Matrix
+import com.mapbox.maps.renderer.gl.GlUtils
+import com.mapbox.maps.renderer.gl.GlUtils.toFloatBuffer
+import java.nio.FloatBuffer
+
+internal class BitmapWidgetRendererImpl(
+  private var bitmap: Bitmap,
+  private val position: WidgetPosition,
+  private val marginX: Float,
+  private val marginY: Float,
+) : WidgetRenderer {
+
+  private var program = 0
+  private var vertexShader = 0
+  private var fragmentShader = 0
+
+  private var attributeVertexPosition = 0
+  private var attributeTexturePosition = 0
+  private var uniformTexture = 0
+  private var uniformMvpMatrix = 0
+  private val textures = intArrayOf(0)
+
+  private var screenMatrix = FloatArray(0)
+  private val rotationMatrix = GlUtils.getIdentityMatrix()
+  private val translateMatrix = GlUtils.getIdentityMatrix()
+
+  private lateinit var mvpMatrixBuffer: FloatBuffer
+  private val mvpMatrix = GlUtils.getIdentityMatrix()
+
+  private var updateMatrix: Boolean = true
+
+  private lateinit var vertexPositionBuffer: FloatBuffer
+
+  private val texturePositionBuffer: FloatBuffer by lazy {
+    floatArrayOf(
+      0f, 0f,
+      0f, 1f,
+      1f, 0f,
+      1f, 1f
+    ).toFloatBuffer()
+  }
+
+  private var bitmapWidth = bitmap.width
```

Comment:
In all cases in the class we use half of width or half of height, should we calculate it only **once** when we assign the value?

Reply:
Bitmap can be updated at https://github.com/mapbox/mapbox-maps-android/pull/1036/files/189635586e4ad09e4610a521c5c38d1a6ccdb0af#diff-cd76e985bf6fa35e9f6b4b18adce883b94d33de6191d2de49287a359c7cc0ddbR251, thus this property will be updated accordingly